.. PyGV documentation master file, created by
   sphinx-quickstart on Thu Mar 18 01:14:05 2021.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

Welcome to PyGV's documentation!
================================

PyGV, short for Python Genome Viewer, is a powerful and user-friendly Python package
designed to empower researchers and genomics enthusiasts with the tools they need to create
publication-ready, highly customizable genomic visualizations seamlessly integrated into their
daily workflows.

Key Features and Advantages
---------------------------

* Flexibility: PyGV allows you to effortlessly configure genomic tracks directly within Python code. Each track is represented as a Python class, making it easy to fine-tune track properties, layouts, and styles to suit your specific research needs. As PyGV is built on top of Matplotlib, you can further harness the familiar syntax used for Matplotlib, giving you the power to customize your genomic visualizations to the fullest extent.
* Interactive Integration: In the spirit of pyGenomeTracks, PyGV excels in interactive integration. PyGV is designed to seamlessly integrate with popular platforms, including Jupyter Notebooks, enhancing the user experience by enabling real-time interactivity and dynamic visualization adjustments, all within the familiar Python environment.
* Rich Genomic Data Support: The package offers support for a wide range of genomic data formats, allowing you to visualize genes, regulatory elements, sequencing alignments, genomic interactions, and other genomics features in the way that best serves your research. See a gallery of all supported tracks [here]().
* Reproducibility and Collaboration: With PyGV, creating and sharing reproducible genomic visualizations becomes
  a straightforward task, essential for research collaboration and documentation.
* Publication-Ready Figures: PyGV is committed to producing publication-ready figures. Our goal is to minimize the additional effort you might otherwise spend adjusting fonts, colors, and other details in software like Adobe Illustrator. With PyGV, you can create polished, high-quality figures directly from your genomic visualizations, saving you time and ensuring that your visualizations are ready to meet the different standards of publication.

Get Started
-----------

.. code-block:: shell

   pip install pygv


Indices and tables
==================

.. toctree::
   :maxdepth: 2

   viewer
   tracks
   auto_examples/index


* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
